### PR TITLE
Enable DestinationPropagation by default.

### PR DESCRIPTION
**Pull request by cjgillot**
~~Based on https://github.com/rust-lang/rust/pull/115291.~~

This PR proposes to enable the destination propagation pass by default.
This pass is meant to reduce the amount of copies present in MIR.

At the same time, this PR removes the `RenameReturnPlace` pass, as it is currently unsound.
`DestinationPropagation` is not limited to `_0`, but does not handle borrowed locals.

### CODE REVIEW

File: compiler/rustc_mir_dataflow/src/points.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    /// Converts a `Location` into a `PointIndex`. O(1).
+    #[inline]
+    pub fn point_from_location(&self, location: Location) -> PointIndex {
+        let Location { block, statement_index } = location;
+        let start_index = self.statements_before_block[block];
+        PointIndex::new(start_index + statement_index)
+    }
+
+    /// Converts a `Location` into a `PointIndex`. O(1).
```
Comment (klensy) [inline compiler/rustc_mir_dataflow/src/points.rs:60]: Wrong copy&paste?

File: compiler/rustc_mir_dataflow/src/points.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+/// When we initially compute liveness, we use an interval matrix storing
+/// liveness ranges for each region-vid.
+pub struct LivenessValues<N: Idx> {
+    elements: Rc<DenseLocationMap>,
+    points: SparseIntervalMatrix<N, PointIndex>,
+}
+
+impl<N: Idx> LivenessValues<N> {
+    /// Creates a new set of "region values" that tracks causal information.
+    /// Each of the regions in num_region_variables will be initialized with an
```
Comment (klensy) [inline compiler/rustc_mir_dataflow/src/points.rs:109]: Variable `num_region_variables` referenced multiple times across codebase, but didn't exist now.

File: compiler/rustc_mir_dataflow/src/points.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        debug!("LivenessValues::add_elements(row={:?}, locations={:?})", row, locations);
+        self.points.union_row(row, locations)
+    }
+
+    /// Adds all the control-flow points to the values for `r`.
+    pub fn add_all_points(&mut self, row: N) {
+        self.points.insert_all_into_row(row);
+    }
+
+    /// Returns `true` if the region `r` contains the given element.
```
Comment (klensy) [inline compiler/rustc_mir_dataflow/src/points.rs:140]:
```suggestion
    /// Returns `true` if the region `row` contains the given element.
```

Comment (klensy) [reply compiler/rustc_mir_dataflow/src/points.rs:140]: Or not?

File: compiler/rustc_mir_dataflow/src/points.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    pub fn contains(&self, row: N, location: Location) -> bool {
+        let index = self.elements.point_from_location(location);
+        self.points.row(row).is_some_and(|r| r.contains(index))
+    }
+
+    /// Returns an iterator of all the elements contained by the region `r`
```
Comment (klensy) [inline compiler/rustc_mir_dataflow/src/points.rs:146]:
```suggestion
    /// Returns an iterator of all the elements contained by the region `row`
```

**Comment by cjgillot** [comment]
@rust-timer build c5a641004bc81b51e8d3d9c11764dca9cf580bd7

**Comment by cjgillot** [comment]
r? wg-mir-opt

**Comment by cjgillot** [comment]
r? compiler

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/115291.
@rustbot blocked

File: compiler/rustc_mir_dataflow/src/points.rs (hunk 2 above)
Comment (klensy) [reply compiler/rustc_mir_dataflow/src/points.rs]: Still here.

**Comment by petrochenkov** [comment]
Perf results are mixed.

This needs to be reviewed by someone who knows the context (are the removed passes subsumed by the enabled pass? why?) and can read mir-opt test diffs.
r? @tmiasko

**Comment by wesleywiser** [comment]
r? mir-opt

**Comment by oli-obk** [comment]
logic wise this is fine. Performance wise it's not ideal, but at least we have a FIXMEs to work on and a better pass to edit. r=me after another rebase

**Comment by ehuss** [comment]
@cjgillot I believe this PR has caused the `Building stage1 codegen backend gcc` step in the PR builder to regress from taking about 45s to taking about 80m. Would it be possible to revert if an immediate fix isn't available.

Discussion at https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/x86_64-gnu-llvm-17.20job.20being.20really.20slow

**Comment by cjgillot** [comment]
Posted https://github.com/rust-lang/rust/pull/125794

**Comment by pnkfelix** [comment]
* was reverted due to injecting [big regression for "Building stage1 codegen backend gcc"](https://github.com/rust-lang/rust/pull/115105#issuecomment-2141019310)
* marking as triaged

@rustbot label: +perf-regression-triaged

**Pull request merged**